pycalphad/io/cs_dat.py

```diff
@@ -1159,6 +1162,999 @@ def parse_cs_dat(instring):
+                        continue
+                    except ValueError:
+                        pass
+            print(f'WARNING: Skipped parameter with order {t_order} and coefficient {coeff_string}')
+
+    return coefficients, extra_parameters, has_extra_parameters
+
+def format_coefficient(coeff):
+    # The formatting is inconsistent, so unfortunately each value range is custom
+    if   coeff == 0:
+        coeff_string = '0.00000000     '
+    elif abs(coeff) < 0.1 or abs(coeff) >= 1e8:
+        coeff_string = f'{coeff*10: .7E}'[:15]
+        if coeff < 0:
+            coeff_string = '-.' + coeff_string[1] + coeff_string[3:] + ' '
+        else:
+            coeff_string = '0.' + coeff_string[1] + coeff_string[3:] + ' '
+    elif abs(coeff) < 1:
+        coeff_string = f'{coeff: .16f}'[:10] + '     '
+        if coeff < 0:
+            coeff_string = ' -' + coeff_string[2:]
+    else:
+        coeff_string = f'{coeff: .16f}'[:10] + '     '
+
+    return coeff_string
+
+def format_coefficient_mag(coeff):
+    # The formatting is inconsistent, so unfortunately each value range is custom
+    if   coeff == 0:
+        coeff_string = '0.000000     '
+    elif abs(coeff) < 0.1 or abs(coeff) >= 1e6:
+        coeff_string = f'{coeff*10: .5E}'[:13]
+        if coeff < 0:
+            coeff_string = '-.' + coeff_string[1] + coeff_string[3:] + ' '
+        else:
+            coeff_string = '0.' + coeff_string[1] + coeff_string[3:] + ' '
+    elif abs(coeff) < 1:
+        coeff_string = f'{coeff: .16f}'[:8] + '     '
+        if coeff < 0:
+            coeff_string = ' -' + coeff_string[2:]
+    else:
+        coeff_string = f'{coeff: .16f}'[:8] + '     '
+
+    return coeff_string
+
+def make_constituent_mapping(constituents, endmember_params):
+    # Match endmembers to constituents they are composed of
+    constituent_mapping = [[] for _ in range(len(constituents))]
+    for endmember in endmember_params:
+        sublattice = 0
+        for speciesList in endmember['constituent_array']:
+            for species in speciesList:
+                constituent_mapping[sublattice].append(constituents[sublattice].index(species.name) + 1)
+                sublattice += 1
+    return constituent_mapping
+
+def iterative_substitution(param,symbols):
+    # Iteratively subsitutes a list of symbols into an equation
+    # Doesn't check for circular dependence
+    subs_param = param
+    requires_substitution = True
+    while requires_substitution:
+        requires_substitution = False
+        param_symbols = [s.name for s in subs_param.free_symbols]
+        # Check if any of the symbols in symbols are still in the parameter
+        common_symbols = [s for s in symbols if s in param_symbols]
+        if common_symbols:
+            # Do another round of substitutions if so
+            subs_param = piecewise_fold(simplify(subs_param.subs(symbols)))
+            requires_substitution = True
+    return subs_param
```

Comment:
I think you can replace this function with `Model.symbol_replace` and eliminate the sympy dependency:
```python
# Maximum number of levels deep we check for symbols that are functions of
# other symbols
_MAX_PARAM_NESTING = 32
def symbol_replace(obj, symbols):
    """
    Substitute values of symbols into 'obj'.
    Parameters
    ----------
    obj : SymEngine object
    symbols : dict mapping symengine.Symbol to SymEngine object
    Returns
    -------
    SymEngine object
    """
    try:
        # Need to do more substitutions to catch symbols that are functions
        # of other symbols
        for iteration in range(_MAX_PARAM_NESTING):
            obj = obj.xreplace(symbols)
            undefs = [x for x in obj.free_symbols if not isinstance(x, v.StateVariable)]
            if len(undefs) == 0:
                break
    except AttributeError:
        # Can't use xreplace on a float
        pass
    return obj
```

Reply:
The calls to `piecewise_fold` and `simplify` here are crucial to getting the equations formatted into simple intervals that can be written to a DAT. I tried switching to `symbol_replace` as you highlighted, but it returns a nested piecewise equation that can't be used.

For more details of what needs to happen, you might look at my recent question on [StackExchange](https://stackoverflow.com/questions/72623272/simplifying-nested-piecewise-equations-in-sympy).